### ISSUE TITLE
Support for different keyword-spacing with optional catch binding syntax

### ISSUE THREAD

**Issue by kinsi55** [eslint/eslint#13278]
**What rule do you want to change?**

keyword-spacing

**Does this change cause the rule to produce more or fewer warnings?**

Fewer, depending on other configuration (Never more)

**How will the change be implemented? (New option, new default behavior, etc.)?**

Probably new option

**Please provide some example code that this change will affect:**

```js
try {
	foo();
} catch(ex) {
	return console.error(ex);
}

// https://tc39.es/proposal-optional-catch-binding/
try {
	foo();
} catch { // This is the root issue
	return console.error("Failed");
}
```

**What does the rule currently do for this code?**

With `"catch": {"after": false}` the first example is, as expected, correct, however when ignoring the thrown error like in the bottom example it still expects you to not have a space after the `catch` thus causing obviously ugly code.

**What will the rule do after it's changed?**

My idea is to add a seperate rule, something like `catchNoBody` which defaults to whatever `catch` is set to but allows you to have a different `keyword-spacing` for catch clauses like the second one

**Are you willing to submit a pull request to implement this change?**

I probably would not know how to properly implement this seeing the need for inheritance so no.

**Comment by anikethsaha** [comment]
Thanks for the issue

currently for `catch` clause, this rule checks spaces around `catch` not around `catch(...)`
so the rule is working as expected . with

`"catch": {"after": false}`

```js
try {
  foo();
} catch {
  // < error here
  return console.error("Failed");
}
```

This behavior is the same for `if`, `for`, and others as well.
this rule doesn't check for any space after `()`, so

you can use this

```js
/*eslint keyword-spacing: ["error", { "overrides": {
  "catch": { "after": false}
} }]*/
try {
  foo();
} catch(ex){
  return console.error(ex);
}
try {
  foo();
} catch{
  // This is the root issue
  return console.error("Failed");
}
```
It will be consistent spacing around `catch` and for `"catch": { "after": true}`
```js
/*eslint keyword-spacing: ["error", { "overrides": {
  "catch": { "after": true}
} }]*/
try {
  foo();
} catch (ex) {
  return console.error(ex);
}
try {
  foo();
} catch {
  // This is the root issue
  return console.error("Failed");
}
```


Let me know if I missed something.

**Comment by kinsi55** [comment]
I'm aware of how the current implementation works, but optional catch binding was ofc not a thing when it was designed. I prefer no space behind most keywords (if, for, catch, ...), however only because of the () section that follows behind it. Before / after the curly braces I want there to be a space - currently the only thing I would be able to do is remove the override for catch like you suggested (And thus have a space between catch and its binding) however that would make my style somewhat inconsistent, hence I had hoped that possibly there could be a way to differentiate between catch with and without a binding in this rule.

**Comment by anikethsaha** [comment]
Maybe I am not getting your proposal, can you give some snippet on how you want this to be changed. It would be helpful. 

Thanks.

**Comment by kinsi55** [comment]
Maybe something like this:

```
"keyword-spacing":         			["error", {
	"overrides": {
		"catch": {"after": false, "afterNoBody": true}
	}
}],
```

With that, both these should not cause an error
```
try {
	// foo
} catch(ex) {
	// bar
}

try {
	// foo
} catch {
	// bar
}
```

**Comment by anikethsaha** [comment]
Thanks for clearing, 

though the rule is working as expected cause it checks `catch` not `catch(..)`

I think it can be done behind an option 👍. 

also, about the option, `afterNoBinding`, IMO new option should be something like `noBinding : boolean` which when set to true will follow/use the config set for `catch` in order to keep these two consistent. and if `false` , do nothing like the current one so that the user can have the control.  


lets hear from others about this.

**Comment by kaicataldo** [comment]
Looks to me like we just missed this when we added support for optional catch binding syntax and that this is actually a bug in the rule. I think that this should be caught by default (i.e. check for a space after the `catch` keyword when there is no binding and after the closing parenthesis when there is one).

That being said, it might be prudent to initially put it behind an option to avoid adding more errors for users so soon after we released v7.

**Comment by anikethsaha** [comment]
I think this should be re-open! cause it is bug and it will be agreed by the [new policy](https://eslint.org/blog/2020/05/changes-to-rules-policies#what-about-bugs) as well

**Comment by mdjermanovic** [comment]
I reopened this thinking it was a bug, but now it looks more like an enhancement to me.

`keyword-spacing` controls spaces between `catch` and `(`, `catch` and `{`.

`space-before-blocks` controls spaces between `)` and `{`, so `keyword-spacing` shouldn't enforce anything there.

With the actual schema, I think it's impossible for `keyword-spacing` to know that user wants `catch(e)`, but at the same time `catch {` instead of `catch{`, so this would most likely need a new option.

(this is somewhat similar to the `else++foo` issue, a difference is that this one doesn't seem to be an edge case)
